src/components/impacts/ImpactForm.jsx

```diff
@@ -3,145 +3,139 @@ import PropTypes from 'prop-types';
 import { connect } from 'react-redux';
 import { dispatch } from 'main';
 import { getCategories } from 'modules/categories';
-import { Form, Input, Button, Select } from 'components/form/Form';
+import { getExternalSources } from 'modules/external-sources';
+import { Form, Input, Button, Select, Textarea } from 'components/form/Form';
 import { Autobind } from 'es-decorators';
 
 import compact from 'lodash/compact';
 import difference from 'lodash/difference';
 
-class ImpactForm extends React.Component {
 
-  constructor(props) {
-    super(props);
+const findInTree = (node, cond) => {
+  if (cond(node)) {
+    return node
+  } else {
+    return _.compact((node.children || []).map(n => findInTree(n, cond)))[0]
+  }
+}
 
-    this.form = {};
-    this.state = {
-      categories: {},
-      // used to add new sources
-      external_sources_index: []
-    };
+const childrenInclude = (node, childId) => {
+  return (node.children || []).map(c => c.id).includes(childId)
+}
 
-    // used to remove sources already added in the database
-    this.external_sources_id = [];
-  }
+const findParent = (roots, childId) => {
+    return findInTree(
+      {children: roots},
+      n => childrenInclude(n, childId)
+    )
+}
 
-  /* lifecycle */
-  componentWillMount() {
-    const { sources, values } = this.props;
-    dispatch(getCategories({ type: 'Impact', tree: true, pageSize: 9999, sort: 'name' }));
+const findId = (roots, id) => {
+    return findInTree(
+      {children: roots},
+      n => n.id === id
+    )
+}
 
-    // for editing...
-    if (sources && Object.keys(values).length) {
-      this.setState({
-        external_sources_index: values.external_sources_index.map(sourceId => sourceId)
-      });
+class ImpactForm extends React.Component {
+
+  constructor(props, ...etc) {
+    super(props, ...etc)
+
+    this.state = {
+      data: this.props.values,
     }
   }
 
-  componentWillReceiveProps(nextProps) {
-    if (nextProps.values && nextProps.values) {
-      const { category_parent_id, category_id } = nextProps.values;
+  componentWillMount() {
+    dispatch(getCategories({
+      type: 'Impact',
+      pageSize: 999,
+      sort: 'name',
+      level: 1,
+      include: ['children']
+    }));
+    dispatch(getExternalSources({
+      pageSize: 999,
+    }));
+  }
 
+  componentWillReceiveProps(newProps) {
+    if (!this.state.data.category && newProps.impactCategories.length) {
       this.setState({
-        categories: {
-          parent: category_parent_id,
-          children: category_id
+        data: {
+          category: newProps.impactCategories[0].children[0]
```

Comment:
We know that to be true, since the if that encloses this test for `newProps.impactCategories.length`. Is there something in particular that worries you here?